Overhaul Quil to LaTeX circuit generation

kilimanjaro opened the pull request:
Description
-----------

This is an overhaul of the previous code generating LaTeX circuits. I've got an example notebook included in the the examples directory. This resolves #1023.

Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] All new and existing tests pass locally and on Semaphore.
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
- [x] Functions and classes have useful sphinx-style docstrings.
- [x] (New Feature) The docs have been updated accordingly.
- [x] (Bugfix) The associated issue is referenced above using
      [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords).
- [x] The [changelog](https://github.com/rigetti/pyquil/blob/master/CHANGELOG.md) is updated,
      including author and PR number (@username, gh-xxx).

kilimanjaro commented:
The previous routine was not very usable, and generated ugly stuff like:
![mystery (1)](https://user-images.githubusercontent.com/1039965/67439826-3d1cfa00-f5ac-11e9-9fab-671a789fb86d.png)
(That's `X 0; CNOT 0 1; H 1`)

Attached is an example of the usage of this (showing basic functionality, showing the use of custom settings, and showing the new `PRAGMA LATEX_GATE_GROUP` feature).
[LaTeX Circuits.pdf](https://github.com/rigetti/pyquil/files/3764862/LaTeX.Circuits.pdf)

notmgsk commented:
The original has a drop-shadow. I dunno if I can approve not having a drop-shadow.

notmgsk commented on pyquil/latex/ipython.py:
```diff
@@ -0,0 +1,42 @@
+    :return: PNG image render of the circuit.
+    :rtype: Image
+    """
+    pdflatex_path = shutil.which('pdflatex')
```
Would be nice if these were configurable. Otherwise note in the docstring which external software is required.

kilimanjaro replied:
I'm opting to not allow too much flexibility, but I did add an explicit reference to the requirements in the docstring (both programs and also needed latex stuff).

notmgsk commented on pyquil/latex/latex_generation.py:
```diff
@@ -32,526 +32,467 @@
+
+
+def TIKZ_LEFT_KET(qubit):
+    return r"\lstick{{\ket{{q_{{{qubit}}}}}}}".format(qubit=qubit)
```
good grief

notmgsk commented on pyquil/latex/latex_generation.py:
```diff
@@ -32,526 +32,467 @@
+                self.index = j + 1
+                return
+
+        raise ValueError("Unable to find PRAGMA {} matching {}".format(PRAGMA_END_GROUP, instr))
```
Missing period, inconsistent with other error msgs.

karalekas commented on pyquil/latex/ipython.py:
```diff
@@ -0,0 +1,51 @@
+from .latex_generation import to_latex
+
+
+def display(circuit: Program, settings=None, **image_options) -> Image:
```
what type is settings

karalekas commented on pyquil/latex/ipython.py:
```diff
@@ -0,0 +1,51 @@
+import tempfile
+from IPython.display import Image
+from pyquil import Program
+from .latex_generation import to_latex
```
use absolute imports

karalekas commented on setup.py:
```diff
@@ -75,6 +75,7 @@
         'requests',
         'networkx>=2.0.0',
         'rpcq>=2.7.2',
+        'ipython', # for pyquil.latex.ipython
```
can we make this optional somehow? I'm happy including it in the examples requirements but the pyquil library doesnt need ipython when not being used in an interactive fashion

karalekas commented on pyquil/tests/test_latex.py:
```diff
@@ -1,22 +1,63 @@
+from pyquil.gates import X, Y, RX, CZ, SWAP, MEASURE, CNOT
+from pyquil.latex import to_latex, DiagramSettings
+
+import pytest
```
lets's move the entire `test_latex.py` file into a `tests` directory in the `latex` module

karalekas commented on pyquil/latex/latex_generation.py:
```diff
@@ -30,528 +30,472 @@
     return text
 
 
+def header(settings):
```
doesnt look like this uses `settings`

karalekas commented on examples/LaTeX Circuits.ipynb:
```diff
@@ -0,0 +1,204 @@
+    }
+   ],
+   "source": [
+    "prog = Program(\"X 0\\n CNOT 0 1\\n H 1\")\n",
```
use `+=` to build up programs -- if its annoying to use `+=` then we should open an issue to fix whatever your pain points are (maybe the pragma and dagger?)

kilimanjaro replied:
I've changed the first two examples to avoid using escape characters etc. Those are built using the gate objects, e.g. `Program(X(0), CNOT(0,1), H(1))`. For the last example I'm going to keep it as a big multiline string -- this is how it came to me (I generated it from the compiler), and I don't think anything would be improved by changing this.

karalekas commented on examples/LaTeX Circuits.ipynb:
```diff
@@ -0,0 +1,204 @@
+   ],
+   "source": [
+    "# extra kwargs are passed are passed straight to `IPython.display.Image`\n",
+    "display(prog, width=300)"
```
this gives me an obscure error:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-6f822dbd0b9c> in <module>
      1 # extra kwargs are passed are passed straight to `IPython.display.Image`
----> 2 display(prog, width=300)

~/code/rigetti/github/pyquil/pyquil/latex/ipython.py in display(circuit, settings, **image_options)
     40         result = subprocess.call([pdflatex_path, "-output-directory", tmpdir, texfile.name])
     41         if result != 0:
---> 42             raise RuntimeError("pdflatex error")
     43 
     44         png = os.path.join(tmpdir, 'diagram.png')

RuntimeError: pdflatex error
```

which I see in my jupyter notebook logs is actually:

```
! Package tikz Error: I did not find the tikz library 'quantikz'. I looked for
files named tikzlibraryquantikz.code.tex and pgflibraryquantikz.code.tex, but n
either could be found in the current texmf trees..
```

karalekas replied:
we should add some good documentation for installing all the requirements for this module

kilimanjaro replied:
Adding sphinx docs, and also adding a bit more care to the error reporting here.

karalekas commented on pyquil/latex/ipython.py:
```diff
@@ -0,0 +1,51 @@
+import os
+import subprocess
+import shutil
+import tempfile
```
lets rename this file to `_ipython.py` to make it "private" and just import the `display` function to the `latex` module API (as you already do)

karalekas commented on examples/LaTeX Circuits.ipynb:
```diff
@@ -0,0 +1,204 @@
+   "outputs": [],
+   "source": [
+    "from pyquil.gates import *\n",
+    "from pyquil.quil import Program\n",
```
rename this notebook to `LaTeXQuilCircuits.ipynb` (no space in name)

karalekas commented:
Also, need to update the changelog -- I think this warrants the "Announcements" section.

karalekas commented:
Also -- do we have a way of aligning measurement (or any gates for that matter)? See the following example where I try to measure both qubits but everything is left justified:

![image](https://user-images.githubusercontent.com/3578739/67826241-1f0e3880-fa89-11e9-8ae6-947b4976ad4d.png)

kilimanjaro commented:
There are two items left.

- Add support for rendering `RESET` instructions as feedback subprograms.
- Add optional support for right-alignment of terminal `MEASURE` instructions (on by default).

These could be part of this PR or could be part of a separate one (I think the latter is fairly easy to implement, but am running out of time for this today).

karalekas commented on requirements.txt:
```diff
@@ -3,6 +3,7 @@ numpy
 antlr4-python3-runtime
 requests
 networkx >= 2.0.0
+ipython # for pyquil.latex.ipython
```
lets put this above `# test deps` in an `# optional latex deps` section

karalekas commented on setup.py:
```diff
@@ -79,6 +79,9 @@
         'immutables==0.6',
     ],
+    extras_require={
+        'display': ["ipython"]
```
I think `latex` might be a better name for the extras group -- thoughts? I get that just the "display" function is what needs it, but "display" alone doesnt really ring a bell for me idk

kilimanjaro replied:
I agree, `latex` is more mnemonic

karalekas commented on pyquil/latex/latex_generation.py:
```diff
@@ -13,545 +13,478 @@
+
+    impute_missing_qubits: bool = False
+    """
+    Include qubits with indices between those explicitly referenced in the Quil program.
```
like we discussed let's put all the Diagram and tikz stuff (so basically line 167 on) into a `_diagram.py` file and export just the `DiagramSettings` object

kilimanjaro replied:
Done

karalekas commented on pyquil/latex/_ipython.py:
```diff
@@ -0,0 +1,68 @@
+import tempfile
+from typing import Optional
+from IPython.display import Image
+
```
copy the license up here

karalekas commented on pyquil/latex/latex_generation.py:
```diff
@@ -13,545 +13,478 @@
+
+import sys
+from copy import copy
+from warnings import warn
```
unused

kilimanjaro replied:
I found a use

karalekas commented:
still need to update changelog, and I think the right alignment should be done for this PR. RESET and classical feedback control can definitely be another PR

kilimanjaro commented:
This should be essentially good to go. I've incorporated feedback, and added support for alignment of terminal measurements e.g. 
![aligned](https://user-images.githubusercontent.com/1039965/67911927-7eca1980-fb45-11e9-8488-a3457945b135.png)
